### PR TITLE
elb facts fixup

### PR DESCRIPTION
**Pull request by nand0p**
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

ec2_elb_facts
##### ANSIBLE VERSION

ansible 1.9.5
##### SUMMARY

updated the ec2_elb_facts mod so as to not error if the elb being queried does not exist.  also returns list of dicts with info for each elb

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @nand0p. To the current maintainers, @mjschultz please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/amazon/ec2_elb_facts.py [hunk 1]
```diff
@@ -72,140 +74,164 @@
...
-            'load_balancer_port': listener[0],
-            'instance_port': listener[1],
-            'protocol': listener[2],
+        if path:
+            health_check_dict['ping_path'] = path
+        return health_check_dict
+
+    def _get_elb_info(self, elb):
+        print "ELB"
+        print elb
```
Comment (mjschultz) [inline cloud/amazon/ec2_elb_facts.py]: Probably don't need these `print` statements anymore.

File: cloud/amazon/ec2_elb_facts.py [hunk 2]
```diff
@@ -72,140 +74,164 @@
...
+            all_elbs = self.connection.get_all_load_balancers()
+        except BotoServerError as err:
+            self.module.fail_json(msg = "%s: %s" % (err.error_code, err.error_message))
+
+        for existing_lb in all_elbs:
+            for lookup_lb in self.names:
+                if lookup_lb in str(existing_lb):
+                    try:
+                        this_lookup = self.connection.get_all_load_balancers(lookup_lb)[0]
+                    except BotoStandardError as err:
```
Comment (mjschultz) [inline cloud/amazon/ec2_elb_facts.py]: BotoStandardError appears to be undefined?

File: cloud/amazon/ec2_elb_facts.py [hunk 3]
```diff
@@ -72,140 +74,164 @@
...
+                        self.module.fail_json(msg=err.message)
 
+                    elb_array.append(self._get_elb_info(this_lookup))
+                    #return self._get_elb_info(this_lookup)
```
Comment (mjschultz) [inline cloud/amazon/ec2_elb_facts.py]: Remove this if it isn't needed please

File: cloud/amazon/ec2_elb_facts.py [hunk 4]
```diff
@@ -72,140 +74,164 @@
...
-    module.exit_json(elbs=elb_array)
+            all_elbs = self.connection.get_all_load_balancers()
+        except BotoServerError as err:
+            self.module.fail_json(msg = "%s: %s" % (err.error_code, err.error_message))
+
+        for existing_lb in all_elbs:
+            for lookup_lb in self.names:
+                if lookup_lb in str(existing_lb):
+                    try:
+                        this_lookup = self.connection.get_all_load_balancers(lookup_lb)[0]
```
Comment (mjschultz) [inline cloud/amazon/ec2_elb_facts.py]: I think the above lines could be simplified to be something like:

```
for existing_lb in all_elbs:
    if existing_lb.name in self.names:
        elb_array.append(self._get_elb_info(existing_lb))
```

perhaps?

**Comment by mjschultz** [comment]
I think it's pretty close, but still needs_revision ; ping me after the revisions.

Also this will end up conflicting with https://github.com/ansible/ansible-modules-extras/pull/2081 I think because of the other changes.

**Comment by gregdek** [comment]
Thanks @nand0p for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by nand0p** [comment]
ready_for_review @mjschultz

**Comment by gregdek** [comment]
Thanks @nand0p. To the current maintainers, @mjschultz please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by nand0p** [comment]
@mjschultz any word on this?

**Comment by mjschultz** [comment]
Hi, sorry, I kept meaning to check this but couldn't get around to it.

So my test playbook for this is something like:

```
- hosts: localhost

  tasks:
    - name: Gather ELB Facts
      local_action:
        module: ec2_elb_facts_new
        region: "us-east-1"
        names:
        - "site-staging-lb"
        - "site-notexist-lb"
        - "site-production-lb"
      register: elb_facts

    - name: all facts
      debug: msg="{{ elb_facts }}"
```

Where `ec2_elb_facts_new` is this module.

When I run this playbook it fails with an exception:

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
[...]
  <Error>
    <Type>Sender</Type>
    <Code>LoadBalancerNotFound</Code>
    <Message>Cannot find Load Balancer site-notexist</Message>
  </Error>
[...]
```

Which I think is what you're trying to avoid based on the

> updated the ec2_elb_facts mod so as to not error if the elb being queried does not exist

In the description? If I'm misinterpreting this let me know.

Also, the tags that end up getting returned seem to be for all load balancers queried instead of just the one that the dict should be describing?

**Comment by nand0p** [comment]
hmm, looks like my code only works when a single elb name/id is specified.  i will update this. thanks.

**Comment by gregdek** [comment]
Thanks @nand0p for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@nand0p A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@nand0p Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by nand0p** [comment]
@mjschultz updated to support multiple elbs.  

also fixes breaking on non-existing elb names being queried.

**Comment by gregdek** [comment]
@nand0p A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by nand0p** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @nand0p. To the current maintainers, @mjschultz please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by mjschultz** [comment]
shipit

Thanks @nand0p, sorry it took so long to get this in!

**Comment by gregdek** [comment]
Thanks again to @nand0p for this PR, and thanks @mjschultz for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request merged**